File: Leetcode/1352. Product of the Last K Numbers.cpp
#include <vector>
using namespace std;

class ProductOfNumbers {
public:
    vector<int> p;
    
    ProductOfNumbers() {
        p.push_back(1);
    }
    
    void add(int num) {
        if (num == 0) {
            p.clear();  // Reset on zero
            p.push_back(1);
        } else {
            p.push_back(p.back() * num);
        }
    }
    
    int getProduct(int k) {
        if (k >= p.size()) {  
            return 0;  // Prevents out-of-bounds access
        }
        return p.back() / p[p.size() - k - 1];
    }
};

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers* obj = new ProductOfNumbers();
 * obj->add(num);
 * int param_2 = obj->getProduct(k);
 */

File: Leetcode/1352. Product of the Last K Numbers.py
class ProductOfNumbers:

    def __init__(self):
        self.p=[1]

    def add(self, num: int) -> None:
        if num==0:
            self.p=[0]*(len(self.p))
            self.p.append(1)
        else:
            self.p.append(self.p[-1]*num)

    def getProduct(self, k: int) -> int:
        #print(self.p)
        if self.p[-k-1]==0:
            return 0
        return self.p[-1]//self.p[-k-1]
        
# Your ProductOfNumbers object will be instantiated and called as such:
# obj = ProductOfNumbers()
# obj.add(num)
# param_2 = obj.getProduct(k)
